finality-aleph/src/data_io/status_provider.rs

```diff
@@ -0,0 +1,397 @@
+            let curr_block = (
+                proposal.branch[i],
+                bottom_num + <NumberFor<B>>::saturated_from(i),
+            )
+                .into();
+            match self.get_parent_hash(&curr_block) {
+                Ok(parent_hash) => {
+                    if parent_hash != proposal.branch[i - 1] {
+                        return false;
+                    }
+                }
+                Err(()) => {
+                    warn!(target: "afa", "A proposal whose top block is available but the whole branch is not {:?}", proposal);
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    fn get_status_inner(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: Option<&ProposalStatus>,
+    ) -> ProposalStatus {
+        use crate::data_io::proposal::AvailableKind::*;
+        use crate::data_io::proposal::ProposalStatus::*;
+
+        if self.is_hopeless_fork(proposal) {
+            return Available(HopelessFork);
+        }
+
+        let old_status = match old_status {
+            Some(status) => status,
+            None => &PendingTopBlock,
+        };
+        match old_status {
+            PendingTopBlock => {
+                let top_block = proposal.top_block();
+                if self.is_block_imported(&top_block) {
+                    if self.is_branch_correct(proposal) {
+                        if self.is_ancestor_finalized(proposal) {
+                            Available(BranchOk)
+                        } else {
+                            // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                            TopBlockImportedButNotFinalizedAncestor
+                        }
+                    } else {
+                        // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                        TopBlockImportedButIncorrectBranch
+                    }
+                } else {
+                    // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                    PendingTopBlock
+                }
+            }
+            TopBlockImportedButNotFinalizedAncestor => {
+                if self.is_ancestor_finalized(proposal) {
+                    Available(BranchOk)
+                } else {
+                    // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                    TopBlockImportedButNotFinalizedAncestor
+                }
+            }
+            TopBlockImportedButIncorrectBranch => {
+                // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                TopBlockImportedButIncorrectBranch
+            }
+            Available(_) => {
+                error!(target: "afa", "Checking availability of an available proposal {:?} status: {:?}", proposal, old_status);
```

Comment:
shouldn't be a `warn!`? the code definitely is doing weird things, but it is not necessarily a breakage

Reply:
Definitely